libass/ass_parse.c

```diff
@@ -120,9 +120,10 @@ void ass_update_font(RenderContext *state)
  */
 static inline int32_t dtoi32(double val)
 {
-    if (isnan(val) || val <= INT32_MIN || val >= INT32_MAX + 1LL)
+    if (val > INT32_MIN && val < INT32_MAX + 1LL)
```

Comment:
For reference, in general using `>`/`<` in float-int comparisons is not safe.
All values will be implicitly cast to floating-point type before comparison; but the precision of the floating -point type may not be sufficient to accurately represent the integer value resulting in implementation-defined rounding to an adjacent represnetable float value. This may lead to the cast integer bounds having a greater magnitude than originally allowing too large values to pass the check. Afterwards casting those too large values to the integer type invokes UB.

However, if my understanding is correct this specific case is still safe. Integer operands are `±2^31 ≈ 2.1475e+09` and `double` is iiuc guaranteed to be able to at least represent 10 decimal digits accurately *(see `DBL_DIG` (and `DECIMAL_DIG`, but the latter doesn't necessarily apply to `double`))*
(And `long long` is guaranteed to cover at least the complement-of-two 64-bit range making the `+ 1LL` safe)

This isn’t obvious though and i had to look up the minimum precision values, so i’d prefer a comment added; e.g.:
```C
// floating-point imprecision not relevant in cast; see DBL_DIG
```

Reply:
Precision doesn’t matter here, as long as the int-to-float conversion produces the closest representable value. If the true value of `(float) INT32_MIN` is equal to `INT32_MIN` or closer to zero, `val > INT32_MIN` trivially implies `(int)val >= INT32_MIN`. If it’s further away from zero, `val > INT32_MIN` still implies that `val` is at least one ulp closer to zero than that, and there’s no representable value that’s closer to zero than `val` but still further away than `INT32_MIN`: if there were such a value, then `(float) INT32_MIN` would have been that value.

Reply:
(I’ve edited the preceding comment to fix some typos and add more clarity.)

You might be confusing `>` with `>=`, although I would still caution that each instance be considered separately (e. g. we have `>=` here before this PR, but the control flow is inverted, so it’s still safe by the same reasoning). Indeed, this is precisely why the conversion path uses strictly-unequal bounds and why the upper bound is written as `< INT32_MAX + 1LL` rather than `<= INT32_MAX` (as mentioned in the previous commit that touched this line, 84928b1fb100004686ac61851720da9247cad34b). If you feel a comment is needed, that’s fair, but I don’t believe `DBL_DIG` is relevant.

Reply:
And that’s enough. As long as it can’t be a more _distant_ value, everything works.

Reply:
i didn't notice the control flow got iverted :facepalm...
nvm then everything is fine, though `DBL_DIG` suggest we even _could_ use `<=` checks here but leaving it as is is probably fine and always obviously *(when not misreading...)* correct